jax/_src/lib/xla_bridge.py

```diff
@@ -124,6 +124,9 @@ def get_compile_options(
       msg = 'device_assignment does not match num_partitions: {} vs {}.'
       raise ValueError(msg.format(device_assignment, num_partitions))
 
+    if device_assignment.dtype == object:
```

Comment:
Could we make this unconditional and require an 2d sequence of objects, fixing up any remaining callers?

Presumably the end state we want to get here is the C++ runtime receives a 2D array of device objects rather than IDs, although that's not necessary for this PR. i.e., everywhere we talk about devices, we do so by object.

Reply:
Sounds good, but note that it would become impossible to get_compile_options without a physical device available.

Reply:
Reflecting on this: I think that's fine. If you don't have physical devices, you should not need a device *assignment* to compile: i.e. which specific devices you want to use. You need some abstraction of what that device topology looks like.